File: frontend/src/api/knowledge.ts
import api from '.'

// ── Types ────────────────────────────────────────────────

export interface DocumentResponse {
  id: string
  status: string
  content_summary: string
  content_length: number
  created_at: string
  updated_at: string
  file_path: string
  chunks_count: number | null
  error_msg: string | null
}

export interface DocumentListResponse {
  items: DocumentResponse[]
  total: number
  page: number
  page_size: number
}

export interface DocumentStatusCounts {
  pending: number
  processing: number
  preprocessed: number
  processed: number
  failed: number
}

export interface KnowledgeQueryResponse {
  answer: string
  mode_used: string
  processing_time_ms: number
}

export interface GraphNode {
  id: string
  label: string
  properties: Record<string, unknown>
}

export interface GraphEdge {
  source: string
  target: string
  relation: string
  properties: Record<string, unknown>
}

export interface SubgraphResponse {
  nodes: GraphNode[]
  edges: GraphEdge[]
}

export interface QueryMode {
  id: string
  name: string
  description: string
}

// ── API ──────────────────────────────────────────────────

export const knowledgeApi = {
  // Document Management
  ingestText: (data: { text: string; description?: string }) =>
    api.post('/knowledge/documents/text', data),

  uploadFile: (file: File) => {
    const formData = new FormData()
    formData.append('file', file)
    return api.post('/knowledge/documents/upload', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
      timeout: 120000,
    })
  },

  listDocuments: (params?: { status?: string; page?: number; page_size?: number }) =>
    api.get<DocumentListResponse>('/knowledge/documents', { params }),

  getDocumentStatus: () =>
    api.get<DocumentStatusCounts>('/knowledge/documents/status'),

  getDocument: (docId: string) =>
    api.get<DocumentResponse>(`/knowledge/documents/${docId}`),

  deleteDocument: (docId: string) =>
    api.delete(`/knowledge/documents/${docId}`),

  // Graph Exploration
  getGraphLabels: () =>
    api.get<string[]>('/knowledge/graph/labels'),

  getSubgraph: (data: { label: string; max_depth?: number; max_nodes?: number }) =>
    api.post<SubgraphResponse>('/knowledge/graph/subgraph', data),

  getEntityInfo: (entityName: string) =>
    api.post('/knowledge/graph/entity/info', { entity_name: entityName }),

  deleteEntity: (entityName: string) =>
    api.delete(`/knowledge/graph/entity/${entityName}`),

  // Query
  query: (data: {
    question: string
    mode?: string
    top_k?: number
    only_need_context?: boolean
    response_type?: string
    conversation_history?: Array<{ role: string; content: string }>
  }) => api.post<KnowledgeQueryResponse>('/knowledge/query', data),

  queryStream: (data: {
    question: string
    mode?: string
    top_k?: number
    response_type?: string
    conversation_history?: Array<{ role: string; content: string }>
  }) => {
    return fetch('/api/v1/knowledge/query/stream', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    })
  },

  getModes: () =>
    api.get<{ modes: QueryMode[] }>('/knowledge/modes'),
}

File: backend/app/schemas/knowledge.py
from pydantic import BaseModel, Field


# ── 文档管理 ──────────────────────────────────────────────

class DocumentIngestText(BaseModel):
    text: str = Field(..., min_length=1)
    description: str | None = None


class DocumentResponse(BaseModel):
    id: str
    status: str
    content_summary: str = ""
    content_length: int = 0
    created_at: str = ""
    updated_at: str = ""
    file_path: str = ""
    chunks_count: int | None = None
    error_msg: str | None = None


class DocumentListResponse(BaseModel):
    items: list[DocumentResponse]
    total: int
    page: int
    page_size: int


class DocumentStatusCounts(BaseModel):
    pending: int = 0
    processing: int = 0
    preprocessed: int = 0
    processed: int = 0
    failed: int = 0


# ── 查询 ──────────────────────────────────────────────────

class KnowledgeQueryRequest(BaseModel):
    question: str = Field(..., min_length=1)
    mode: str = Field("hybrid", pattern="^(local|global|hybrid|mix|naive|bypass)$")
    top_k: int = Field(10, ge=1, le=200)
    only_need_context: bool = False
    response_type: str = "Multiple Paragraphs"
    conversation_history: list[dict[str, str]] = []
    stream: bool = False


class KnowledgeQueryResponse(BaseModel):
    answer: str
    mode_used: str
    processing_time_ms: int


class QueryModeInfo(BaseModel):
    id: str
    name: str
    description: str


# ── 图谱 ──────────────────────────────────────────────────

class GraphNode(BaseModel):
    id: str
    label: str
    properties: dict = {}


class GraphEdge(BaseModel):
    source: str
    target: str
    relation: str
    properties: dict = {}


class SubgraphRequest(BaseModel):
    label: str
    max_depth: int = Field(3, ge=1, le=10)
    max_nodes: int = Field(100, ge=1, le=1000)


class SubgraphResponse(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]


class EntityInfoRequest(BaseModel):
    entity_name: str


class EntityUpdateRequest(BaseModel):
    entity_name: str
    updated_data: dict[str, str]
    allow_rename: bool = True
    allow_merge: bool = False


class EntityCreateRequest(BaseModel):
    entity_name: str
    entity_data: dict[str, str]


class EntityMergeRequest(BaseModel):
    source_entities: list[str] = Field(..., min_length=2)
    target_entity: str
    merge_strategy: dict[str, str] | None = None
    target_entity_data: dict[str, str] | None = None


class RelationDeleteRequest(BaseModel):
    source_entity: str
    target_entity: str


class RelationInfoRequest(BaseModel):
    source_entity: str
    target_entity: str


class RelationCreateRequest(BaseModel):
    source_entity: str
    target_entity: str
    relation_data: dict[str, str]


class RelationUpdateRequest(BaseModel):
    source_entity: str
    target_entity: str
    updated_data: dict[str, str]
